src/lib/VirtualScroller.js

```diff
@@ -0,0 +1,183 @@
+class VirtualScroller {
+    /**
+     * [constructor]
+     *
+     * @param {HTMLElement} container - The HTMLElement that will contain the virtual scroller
+     * @return {VirtualScroller} Instance of VirtualScroller
+     */
+    constructor(container) {
+        this.container = container;
+
+        this.previousScrollTop = 0;
+
+        this.onScrollHandler = this.onScrollHandler.bind(this);
+        this.renderItems = this.renderItems.bind(this);
+    }
+
+    /**
+     * Destroys the virtual scroller
+     *
+     * @return {void}
+     */
+    destroy() {
+        if (this.containerEl) {
+            this.containerEl.remove();
+        }
+
+        this.containerEl = null;
+        this.contentEl = null;
+    }
+
+    /**
+     * Initializes the virtual scroller
+     *
+     * @param {Object} config - The config
+     * @return {void}
+     */
+    init(config) {
+        if (!config.totalItems || !config.itemHeight) {
+            throw new Error('Need to provide totalItems and itemHeight');
+        }
+
+        // The total number items to be scrolled
+        this.totalItems = config.totalItems;
+
+        // The height of each individual item
+        this.itemHeight = config.itemHeight;
+
+        // The height of the visible container
+        this.containerHeight = config.containerHeight;
+
+        // The callback function that to allow users generate the item
+        this.renderItemFn = config.renderItemFn;
+
+        // Allows the user to specify the margin at the top of the container before the first item is rendered
+        this.marginTop = config.marginTop || 0;
+
+        // Allows the user to specify the margin at the bottom of the container after the last item is rendered
+        this.marginBottom = config.marginBottom || 0;
+
+        // The number of items that can fit in view
+        this.totalViewItems = Math.ceil(this.containerHeight / this.itemHeight);
+
+        // The height of the buffer before virtual scroll renders the next set
+        this.maxBufferHeight = this.totalViewItems * this.itemHeight;
+
+        // The max number of items to render at any one given time
+        this.maxRenderedItems = Math.ceil(this.totalViewItems * 3);
```

Comment:
Can you add a variable somewhere for the buffer multiple, just to clarify its intention?